docs/proposals/004-endpoint-picker-protocol/README.md

```diff
@@ -9,27 +9,53 @@ This doc defines the protocol between the EPP and the proxy (e.g, Envoy).
 The EPP MUST implement the Envoy
 [external processing service](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ext_proc/v3/external_processor) protocol.
 
+## Endpoint Subset
+For each HTTP request, the proxy CAN communicate the subset of endpoints the EPP MUST pick from by setting an unstructured entry in the [filter metadata](https://github.com/envoyproxy/go-control-plane/blob/63a55395d7a39a8d43dcc7acc3d05e4cae7eb7a2/envoy/config/core/v3/base.pb.go#L819) field of the ext-proc request. The metadata entry for the subset list MUST be wrapped with an outer key (which represents the metadata namespace) with a default of `envoy.lb.subset_hint`.
+
+```go
+filterMetadata: {
+  "envoy.lb.subset_hint" {
+     "x-gateway-destination-endpoint-subset-hint": [<ip:port>, <ip:port>, ...]
+  }
+}
+```
+
+If the key `x-gateway-destination-endpoint-subset-hint` is set, the EPP MUST only select endpoints from the specified list. If none of the endpoints in the list is eligible or the list is empty, then the EPP MUST return a 429 status code.
+
+If the key `x-gateway-destination-endpoint-subset-hint` is not set, then the EPP MUST select from the set defined by the `InferencePool` selector.
+
+## Destination Endpoint
 For each HTTP request, the EPP MUST communicate to the proxy the picked model server endpoint via:
 
 1. Setting the `x-gateway-destination-endpoint` HTTP header to the selected endpoint in <ip:port> format.
 
-2. Set an unstructured entry in the [dynamic_metadata](https://github.com/envoyproxy/go-control-plane/blob/c19bf63a811c90bf9e02f8e0dc1dcef94931ebb4/envoy/service/ext_proc/v3/external_processor.pb.go#L320) field of the ext-proc response. The metadata entry for the picked endpoint MUST be wrapped with an outer key (which represents the metadata namespace) with a default of `envoy.lb`.
+2. Set an unstructured entry in the [dynamic_metadata](https://github.com/envoyproxy/go-control-plane/blob/c19bf63a811c90bf9e02f8e0dc1dcef94931ebb4/envoy/service/ext_proc/v3/external_processor.pb.go#L320) field of the ext-proc response. The metadata entry for the picked endpoints MUST be wrapped with an outer key (which represents the metadata namespace) with a default of `envoy.lb`.
 
-The final metadata necessary would look like: 
+The primary endpoint MUST be set using the key `x-gateway-destination-endpoint` as follows:
 ```go
 dynamicMetadata: {
   "envoy.lb": {
-    "x-gateway-destination-endpoint": <ip:port>"  
+    "x-gateway-destination-endpoint": <ip:port>
   }
 }
 ```
 
-Note:
+Fallback endpoints MUST be set using the key `x-gateway-destination-endpoint-fallbacks` as a list as follows:
```

Comment:
Ideally we should have a section about `x-gateway-destination-endpoint-fallbacks` and its purpose/expectation above.

Reply:
done